File: backend/app/schemas/restaurant.py
"""Pydantic schemas for Restaurant CRUD."""

from datetime import datetime, time

from pydantic import BaseModel, ConfigDict, Field, computed_field


class RestaurantCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    address: str | None = Field(None, max_length=300)
    city: str | None = Field(None, max_length=100)
    phone_number: str | None = Field(None, max_length=30)
    image: str | None = Field(None, max_length=500)
    opening_time: time | None = None
    closing_time: time | None = None
    is_open: bool = True


class RestaurantUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    address: str | None = Field(None, max_length=300)
    city: str | None = Field(None, max_length=100)
    phone_number: str | None = Field(None, max_length=30)
    image: str | None = Field(None, max_length=500)
    opening_time: time | None = None
    closing_time: time | None = None
    is_open: bool | None = None


class RestaurantResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    owner_id: int
    name: str
    description: str | None = None
    address: str | None = None
    city: str | None = None
    phone_number: str | None = None
    image: str | None = None
    opening_time: time | None = None
    closing_time: time | None = None
    is_open: bool
    average_rating: float = 0.0
    total_reviews: int = 0
    approval_status: str = "approved"
    rejection_reason: str | None = None
    created_at: datetime | None = None

    @computed_field
    @property
    def rating(self) -> float:
        """Backward-compatible alias for average_rating."""
        return self.average_rating


class RestaurantApprovalUpdate(BaseModel):
    approval_status: str = Field(..., pattern="^(approved|rejected)$")
    rejection_reason: str | None = Field(None, max_length=500)


class PopularDishStat(BaseModel):
    dish_id: int
    dish_name: str
    order_count: int


class RestaurantDashboardResponse(BaseModel):
    restaurant_id: int
    restaurant_name: str
    approval_status: str
    total_dishes: int
    available_dishes: int
    average_rating: float
    total_reviews: int
    total_orders: int
    popular_dishes: list[PopularDishStat]

File: frontend/lib/models/restaurant.dart
import 'json_parse.dart';

/// Restaurant from `RestaurantResponse`.
class Restaurant {
  const Restaurant({
    required this.id,
    required this.name,
    this.ownerId,
    this.description,
    this.address,
    this.city,
    this.phoneNumber,
    this.image,
    this.isOpen = true,
    this.averageRating = 0,
    this.totalReviews = 0,
    this.approvalStatus = 'approved',
    this.rejectionReason,
    this.tags = const [],
    this.openingTime,
    this.closingTime,
    this.createdAt,
  });

  final int id;
  final String name;
  final int? ownerId;
  final String? description;
  final String? address;
  final String? city;
  final String? phoneNumber;
  final String? image;
  final bool isOpen;
  final double averageRating;
  final int totalReviews;
  final String approvalStatus;
  final String? rejectionReason;
  final List<String> tags;
  final String? openingTime;
  final String? closingTime;
  final DateTime? createdAt;

  factory Restaurant.fromJson(Map<String, dynamic> json) => Restaurant(
        id: parseInt(json['id'], field: 'id'),
        name: parseString(json['name']),
        ownerId: parseIntOrNull(json['owner_id']),
        description: json['description']?.toString(),
        address: json['address']?.toString(),
        city: json['city']?.toString(),
        phoneNumber: json['phone_number']?.toString(),
        image: json['image']?.toString(),
        isOpen: parseBool(json['is_open']),
        averageRating: parseDoubleOrNull(json['average_rating']) ??
            parseDoubleOrNull(json['rating']) ??
            0,
        totalReviews: parseIntOrNull(json['total_reviews']) ?? 0,
        approvalStatus: json['approval_status']?.toString() ?? 'approved',
        rejectionReason: json['rejection_reason']?.toString(),
        tags: _parseTags(json['tags']),
        openingTime: json['opening_time']?.toString(),
        closingTime: json['closing_time']?.toString(),
        createdAt: parseDateTimeOrNull(json['created_at']),
      );

  static List<String> _parseTags(dynamic value) {
    if (value is List) {
      return value.map((e) => e.toString()).where((t) => t.isNotEmpty).toList();
    }
    return const [];
  }
}
